cmd/src/main.go

```diff
@@ -96,23 +99,39 @@ func readConfig() (*config, error) {
 		}
 	}
 
+	envToken := os.Getenv("SRC_ACCESS_TOKEN")
+	envEndpoint := os.Getenv("SRC_ENDPOINT")
+
+	if userSpecified {
+		// If a config file is present, either zero or both environment variables must be present.
+		// We don't want to partially apply environment variables.
+		if envToken == "" && envEndpoint != "" {
+			return nil, errConfigMerge
+		}
+		if envToken != "" && envEndpoint == "" {
+			return nil, errConfigMerge
+		}
+	}
+
 	// Apply config overrides.
-	if envToken := os.Getenv("SRC_ACCESS_TOKEN"); envToken != "" {
+	if envToken != "" {
 		cfg.AccessToken = envToken
 	}
-	if *endpoint != "" {
-		cfg.Endpoint = *endpoint
-	}
-	if cfg.Endpoint == "" {
-		if endpoint := os.Getenv("SRC_ENDPOINT"); endpoint != "" {
-			cfg.Endpoint = endpoint
-		}
+	if envEndpoint != "" {
+		cfg.Endpoint = envEndpoint
 	}
 	if cfg.Endpoint == "" {
 		cfg.Endpoint = "https://sourcegraph.com"
 	}
 
+	// Lastly, apply endpoint flag if set
+	if endpoint != nil && *endpoint != "" {
+		cfg.Endpoint = *endpoint
+	}
+
 	cfg.Endpoint = strings.TrimSuffix(cfg.Endpoint, "/")
 
 	return &cfg, nil
 }
+
+var errConfigMerge = errors.New("config merging not supported, zero or both environment variables must be set")
```

Comment:
I'd make this more descriptive. As a user it's hard to know what "config merging not supported" means when I just set an env var.

```suggestion
var errConfigMerge = errors.New("when using a configuration file, zero or all environment variables must be set")
```

Or something like that.